### DISCUSSION TITLE
Snapshot tests and reference images

### DISCUSSION THREAD

**Discussion by noahsmartin** [MobileNativeFoundation/discussions#189] [General]
Hi! I'm curious how people are handling snapshot tests for their apps. I've heard of two common ways this is implemented:

1 - Base commit comparisons, where snapshots are generated from the merge base of your pull request and compared against the snapshots generated from the last commit of your pull request. This doesn't require any specific reference images, they are dynamically generated from the base commit. If you change some UI, you have to accept the changes (usually through a web UI tied to the status check) but you don't commit any changes indicating the reference image changed. 

2 - Reference images in the repo. This is when snapshot images are checked in to the repo and used to compare images generated from your commit. If you change any UI you have to replace the image in the repo with a new image.

I'm only familiar firsthand with the first approach, but I've heard about companies using the second. How are others handling this, and which is preferred?

**Comment by shahzadmajeed** [comment]
We are using approach # 2 at DraftKings. Which means all snapshots are pre-recorded and compared against latest commit. There are 2 things to keep in mind: 
- This approach can dramatically increase your repo size so you might want to consider Git LFS or some external storage. 
- You have to be very careful about the target membership of snapshot images. You want to make sure that none of them are added to any of your app target. A pre-commit hook can do the job.

This approach is time saving in terms of build times as you are only creating one snapshot of the view unlike # 1 where you have to generate base snapshot from parent branch.

Another thing to consider is that you should be able to instantiate and snapshot test themed views for each supported theme in-parallel for performance gain.

**Reply by noahsmartin** [reply]
Good to know, thanks!

**Comment by codeman9** [comment]
At Lyft we are using approach #2. We keep reference images stored out-of-repo (Artifactory) and keep a json file containing various SHAs within the modules of our repo referencing which images to use/download for comparison. Through our bazel configuration, images for any particular module are downloaded to use on-demand. If images need to be updated, a comment is left on a PR to accept the new changes and CI auto-updates the reference images. During the PR process, any snapshot failure images that are generated are uploaded and referenced to show directly on the GitHub Checks tab.
